abi/realloc.go

```diff
@@ -0,0 +1,52 @@
+package abi
+
+import "unsafe"
+
+// realloc allocates or reallocates memory for Component Model calls across
+// the host-guest boundary.
+//
+// Note: the use of uintptr assumes 32-bit pointers, e.g. GOOS=wasm32 when compiled for WebAssembly.
+//
+//go:export cabi_realloc
+//go:wasmexport cabi_realloc
+func realloc(ptr unsafe.Pointer, size, align, newsize uintptr) unsafe.Pointer {
+	if newsize <= size {
+		return unsafe.Add(ptr, offset(uintptr(ptr), align))
```

Comment:
I understand that when `newsize <= size`, it returns the original pointer with additional offset. But is this really working without reallocating and copying the old memory block to a new memory block?

Reply:
That’s an interesting example. If you realloc with a different alignment than the original allocation, presumably you’re storing a different type, in which case the copy shouldn’t happen. What does the spec say here?

Reply:
Good idea.

https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#canonical-abi

> The (realloc ...) option specifies a core function that is validated to have the following core function type:

```wat
(func (param $originalPtr i32)
      (param $originalSize i32)
      (param $alignment i32)
      (param $newSize i32)
      (result i32))
```

> The Canonical ABI will use realloc both to allocate (passing 0 for the first two parameters) and reallocate. If the Canonical ABI needs realloc, validation requires this option to be present (there is no default).